tests/test_functional.py

```diff
@@ -0,0 +1,403 @@
+
+        expected_messages, expected_text = self._get_expected()
+        received_messages, received_text = self._get_received()
+
+        if expected_messages != received_messages:
+            msg = ['Wrong results for file "%s":' % (self._test_file.base)]
+            missing, unexpected = multiset_difference(expected_messages,
+                                                      received_messages)
+            if missing:
+                msg.append('\nExpected in testdata:')
+                msg.extend(' %3d: %s' % msg for msg in sorted(missing))
+            if unexpected:
+                msg.append('\nUnexpected in testdata:')
+                msg.extend(' %3d: %s' % msg for msg in sorted(unexpected))
+            pytest.fail('\n'.join(msg))
+        self._check_output_text(expected_messages, expected_text, received_text)
+
+    def _split_lines(self, expected_messages, lines):
+        emitted, omitted = [], []
+        for msg in lines:
+            if (msg[1], msg[0]) in expected_messages:
+                emitted.append(msg)
+            else:
+                omitted.append(msg)
+        return emitted, omitted
+
+    def _check_output_text(self, expected_messages, expected_lines,
+                           received_lines):
+        assert self._split_lines(expected_messages, expected_lines)[0] == \
+            received_lines
+
+
+class LintModuleOutputUpdate(LintModuleTest):
+
+    def _open_expected_file(self):
+        try:
+            return super(LintModuleOutputUpdate, self)._open_expected_file()
+        except IOError:
+            return io.StringIO()
+
+    def _check_output_text(self, expected_messages, expected_lines,
+                           received_lines):
+        if not expected_messages:
+            return
+        emitted, remaining = self._split_lines(expected_messages, expected_lines)
+        if emitted != received_lines:
+            remaining.extend(received_lines)
+            remaining.sort(key=lambda m: (m[1], m[0], m[3]))
+            with open(self._test_file.expected_output, 'w') as fobj:
+                writer = csv.writer(fobj, dialect='test')
+                for line in remaining:
+                    writer.writerow(line.to_csv())
+
+
+def get_tests():
+    input_dir = os.path.join(os.path.dirname(os.path.abspath(__file__)),
+                             'functional')
+    suite = []
+    for fname in os.listdir(input_dir):
+        if fname != '__init__.py' and fname.endswith('.py'):
+            suite.append(FunctionalTestFile(input_dir, fname))
+    return suite
+
+
+TESTS = get_tests()
+TESTS_NAMES = [t.base for t in TESTS]
+
+
+def test_a_functional_test_exists_for_all_rule_names():
+    tests = set(TESTS_NAMES)
+    for rule_name in google_styleguide.GoogleStyleGuideChecker.msgs.values():
+        test_name = rule_name[1].replace('-', '_')
+        assert (test_name in tests) or ({test_name + '2', test_name + '3'}.issubset(tests))
```

Comment:
this tests that a functional test file must exist for every rule added to `GoogleStyleGuideChecker.msgs`